### DISCUSSION TITLE
Span.currentSpan() returns a span whose Trace's id is 00000000000000000000000000000000 instead of the actual current trace's id

### DISCUSSION THREAD

**Discussion by yatamraghu7** [open-telemetry/opentelemetry-java-instrumentation#12217] [Q&A] [answered]
Hi All,

As provided a resolution here https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2543
We are using mvn dependency approach for instrumenting our spring boot with opentelemetry.
We are getting the issue when we use Span currentSpan = Span.current();
PropagatedSpan{ImmutableSpanContext{traceId=00000000000000000000000000000000, spanId=0000000000000000, traceFlags=00, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=false}}
Can you please tell us what can we change to get the proper traceId and spanId which was supposed to get instrumented.
We are not using agent in our project and also no java code has been written for instrumentation. Below is the bom that we are using 

```
<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>io.opentelemetry.instrumentation</groupId>
				<artifactId>opentelemetry-instrumentation-bom</artifactId>
				<version>2.6.0</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
</dependencyManagement>
```

```
              <dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-starter-sleuth</artifactId>
			<version>3.1.3</version>
			<exclusions>
				<exclusion>
					<groupId>org.springframework.cloud</groupId>
					<artifactId>spring-cloud-sleuth-brave</artifactId>
				</exclusion>
			</exclusions>
		</dependency>
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-sleuth-otel-autoconfigure</artifactId>
			<version>1.1.2</version>
		</dependency>
                 <dependency>
			<groupId>io.opentelemetry</groupId>
			<artifactId>opentelemetry-exporter-otlp</artifactId>
		</dependency>
		<dependency>
			<groupId>io.opentelemetry.instrumentation</groupId>
			<artifactId>opentelemetry-spring-boot-starter</artifactId>
		</dependency>
```

Thanks

**Comment by laurit** [comment]
A trace/span id of all zeros denotes an invalid trace, this is used when there is no active trace. Without knowing anything about your application it is hard to tell why this happens. It would be best if you could provide a minimal application that reproduces your issue along with any necessary instructions. You may also try with the agent and see if it works any better.

**Reply by yatamraghu7** [reply]
Hi @laurit 
Can you please let me know where can I find the agent compatible for my opentelemetry version that I am using which is 2.6.0
I would like to give a try with agent and if not working then I will provide a code snippet.

Thanks

**Reply by laurit** [reply]
Have a look at https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases

**Reply by yatamraghu7** [reply]
Hi @laurit 

I tried using the java agent but still the same issue. Here is my problem in detail below.

Issue:

We are intercepting the our httpRequests using sleuth TracingClientHttpRequestInterceptor and below is the code snippet that I am using 

```
@Configuration
public class SleuthConfiguration {

    @Autowired
    private CurrentTraceContext currentTraceContext;

    @Autowired
    private HttpClientHandler httpClientHandler;

    @Bean
    public RestTemplate restTemplate() {
        return new RestTemplateBuilder()
                .interceptors(TracingClientHttpRequestInterceptor.create(currentTraceContext, httpClientHandler))
            .build();
    }
}
```
As per my dependencies mentioned above I am excluding the brave trace propagators from sleuth and somehow our tracing interceptors are mixed up with opentelemetry. So we decided to get rid of sleuth and implement opentelemetry and when I am trying to achieve that I am having issue when trying to intercept our httpRequests using opentelemetry instead of sleuth which was provided above and below is the code snippet that I am using and dependencies

```
                 <dependency>
			<groupId>io.opentelemetry.instrumentation</groupId>
			<artifactId>opentelemetry-spring-boot-starter</artifactId>
			<version>2.7.0</version>
		</dependency>
		<dependency>
			<groupId>io.opentelemetry.instrumentation</groupId>
			<artifactId>opentelemetry-instrumentation-api</artifactId>
			<version>1.19.0</version>
		</dependency>
```

```
@Configuration
public class SleuthConfiguration {
    @Bean
    public RestTemplate restTemplate() {
        return new RestTemplateBuilder().additionalInterceptors(new RestTemplateInterceptor()).build();
    }
}
```
I am getting error that RestTemplateInterceptor cannot be resolved to a type as the dependecy was missing for this and if i am trying to add below dependency I am getting error 
```

                <dependency>
			<groupId>io.opentelemetry.instrumentation</groupId>
			<artifactId>opentelemetry-resttemplate-1.5</artifactId>
			<version>1.19.0</version>
		</dependency>
```

How can i intercept propertly my httpRequests using opentelemetry ?

Thanks

**Reply by laurit** [reply]
Firstly sleuth and opentelemetry are competing tracing solutions. Sleuth questions are best asked from the sleuth community. If you decide to use open telemetry then a sample application would be handy. Currently it is quite unclear what you are attempting to do.

**Comment by yatamraghu7** [comment]
Hi @laurit 

Here is a sample project and I wanted to see a traceId and spanId in my logs when two of my microservices are communicating with restTemplate.
https://github.com/yatamraghu7/opentelemetryInstrumentation/tree/main
Currently there are no traceId and spanid generated. Can you tell me where I was doing wrong. I have followed this 
https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/spring/spring-web/spring-web-3.1/library

I wanted to do manual instrumentation with opentelemetry because the opentelemetry java agent jar will support RestTemplate instrumentation but not the WebClinet and WebSocket communication between my microservices which I already have currently in my main project.
So it would be great if you can point me to a proper documentation where I can get the information of manual instrumentation of opentelemetry for RestTemplate, WebClient and WebSocket communications.

Thanks

**Reply by laurit** [reply]
Firstly I tried running your sample app with the agent. I deleted `opentelemetry-spring-boot-starter` and `opentelemetry-spring-web-3.1` dependencies from https://github.com/yatamraghu7/opentelemetryInstrumentation/blob/4505f89ff5ecfc2c9c564c0280b98d09307119ae/Consumer/pom.xml#L53 as these are not needed with the agent. Also in https://github.com/yatamraghu7/opentelemetryInstrumentation/blob/4505f89ff5ecfc2c9c564c0280b98d09307119ae/Consumer/src/main/java/com/microservice/consumer/Consumer/Config/RestTemplateConfig.java#L14 I changed the method to
```java
@Bean
RestTemplate restTemplate() {
  return new RestTemplate();
}
```
Then I built the applications and ran with `java -javaagent:/path/to/agent -Dotel.traces.exporter=console -Dotel.metrics.exporter=none -Dotel.logs.exporter=none -jar target/Producer-0.0.1-SNAPSHOT.jar` and `java -javaagent:/path/to/agent -Dotel.traces.exporter=console -Dotel.metrics.exporter=none -Dotel.logs.exporter=none -jar target/Consumer-0.0.1-SNAPSHOT.jar` and hit the url `http://localhost:8080/consume` In the consumer application output I see
```
[otel.javaagent 2024-09-12 16:24:53:080 +0300] [http-nio-8080-exec-1] INFO io.opentelemetry.exporter.logging.LoggingSpanExporter - 'GET' : 3d2fb66bf84c920fdee0cd6b366b0bb9 15f13e51f0770759 CLIENT [tracer: io.opentelemetry.http-url-connection:2.8.0-alpha-SNAPSHOT] AttributesMap{data={server.address=localhost, http.response.status_code=200, thread.id=24, http.request.method=GET, server.port=8081, network.protocol.version=1.1, thread.name=http-nio-8080-exec-1, url.full=http://localhost:8081/hello}, capacity=128, totalAddedValues=8}
[otel.javaagent 2024-09-12 16:24:53:092 +0300] [http-nio-8080-exec-1] INFO io.opentelemetry.exporter.logging.LoggingSpanExporter - 'GET /consume' : 3d2fb66bf84c920fdee0cd6b366b0bb9 c07328a2ab3c0ba6 SERVER [tracer: io.opentelemetry.tomcat-7.0:2.8.0-alpha-SNAPSHOT] AttributesMap{data={server.address=localhost, client.address=0:0:0:0:0:0:0:1, http.response.status_code=200, network.peer.address=0:0:0:0:0:0:0:1, thread.id=24, url.path=/consume, http.request.method=GET, server.port=8080, http.route=/consume, user_agent.original=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36, network.peer.port=63285, network.protocol.version=1.1, url.scheme=http, thread.name=http-nio-8080-exec-1}, capacity=128, totalAddedValues=14}
```
and in the producer
```
[otel.javaagent 2024-09-12 16:24:53:083 +0300] [http-nio-8081-exec-1] INFO io.opentelemetry.exporter.logging.LoggingSpanExporter - 'GET /hello' : 3d2fb66bf84c920fdee0cd6b366b0bb9 4b68068d57cc8e8f SERVER [tracer: io.opentelemetry.tomcat-7.0:2.8.0-alpha-SNAPSHOT] AttributesMap{data={server.address=localhost, client.address=127.0.0.1, http.response.status_code=200, network.peer.address=127.0.0.1, thread.id=25, url.path=/hello, http.request.method=GET, server.port=8081, http.route=/hello, user_agent.original=Java/17.0.11, network.peer.port=63288, network.protocol.version=1.1, url.scheme=http, thread.name=http-nio-8081-exec-1}, capacity=128, totalAddedValues=14}
```
So seems to be working nicely.
Next I tried using the spring boot starter. Spring boot starter is similar to the agent, it configures instrumentation automatically so you don't need to do much. I added back the `opentelemetry-spring-boot-starter` dependency to the consumer pom. I didn't add `opentelemetry-spring-web-3.1` because the starter already depends on it, no need to add it again. Rebuilt and ran the apps with `java -Dotel.traces.exporter=console -Dotel.metrics.exporter=none -Dotel.logs.exporter=none -jar target/Producer-0.0.1-SNAPSHOT.jar` and `java -Dotel.traces.exporter=console -Dotel.metrics.exporter=none -Dotel.logs.exporter=none -jar target/Consumer-0.0.1-SNAPSHOT.jar` and hit the url `http://localhost:8080/consume`. In the consumer application output I see
```
2024-09-12 16:35:20.699  INFO 22398 --- [nio-8080-exec-1] i.o.e.logging.LoggingSpanExporter        : 'GET' : 765c3a10e6411ba65b6d00b791500a7a 7fbed9619a5e9674 CLIENT [tracer: io.opentelemetry.spring-web-3.1:2.7.0-alpha] AttributesMap{data={http.response.status_code=200, url.full=http://localhost:8081/hello, server.port=8081, server.address=localhost, http.request.method=GET}, capacity=128, totalAddedValues=5}
2024-09-12 16:35:20.706  INFO 22398 --- [nio-8080-exec-1] i.o.e.logging.LoggingSpanExporter        : 'GET /consume' : 765c3a10e6411ba65b6d00b791500a7a a9f7494547abe3c6 SERVER [tracer: io.opentelemetry.spring-webmvc-5.3:2.7.0-alpha] AttributesMap{data={http.response.status_code=200, url.scheme=http, server.port=8080, network.peer.port=63461, network.peer.address=0:0:0:0:0:0:0:1, server.address=localhost, client.address=0:0:0:0:0:0:0:1, user_agent.original=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36, url.path=/consume, http.request.method=GET, http.route=/consume, network.protocol.version=1.1}, capacity=128, totalAddedValues=12}
```
and in the producer
```
2024-09-12 16:35:20.697  INFO 22372 --- [nio-8081-exec-1] i.o.e.logging.LoggingSpanExporter        : 'GET /hello' : 765c3a10e6411ba65b6d00b791500a7a f05dc96fa5863bf2 SERVER [tracer: io.opentelemetry.spring-webmvc-5.3:2.7.0-alpha] AttributesMap{data={http.response.status_code=200, url.scheme=http, server.port=8081, network.peer.port=63463, network.peer.address=127.0.0.1, server.address=localhost, client.address=127.0.0.1, user_agent.original=Java/17.0.11, url.path=/hello, http.request.method=GET, http.route=/hello, network.protocol.version=1.1}, capacity=128, totalAddedValues=12}
```
so this also works fine.

> I wanted to do manual instrumentation with opentelemetry because the opentelemetry java agent jar will support RestTemplate instrumentation but not the WebClinet and WebSocket communication

Agent also supports the spring WebClient. It is possible to use the agent along with manual instrumentation.

> So it would be great if you can point me to a proper documentation where I can get the information of manual instrumentation of opentelemetry for RestTemplate, WebClient and WebSocket communications.

If you wish to do it manually without using the agent or spring starter then 
rest template - https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/spring/spring-web/spring-web-3.1/library 
web client - https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-webflux/spring-webflux-5.3/library/README.md
websocket - not supported, you'll have to figure this out on your own
you should be aware that rest template and web client are just http client you'll probably also need server instrumentation

**Reply by yatamraghu7** [reply]
Hi @laurit 

Thanks a lot for detailed analysis and here are my further questions below

1. How can I enable X-B3 trace headers with open-telemetry. Do i need to make some code change in my java somewhere ?
2. Will agent support Web Socket ? Any idea when this can be supported with Java Agent ? Can you please give me some starting point where I can achieve the web socket with manual instrumentation at least.

Thanks

**Reply by laurit** [reply]
1. See https://opentelemetry.io/docs/languages/java/configuration/#propagators you will need to configure the agent to use a b3 propagator (either `b3` or `b3multi`). It is possible to use multiple propagators at the same time.
2. Context propagation is handled by client passing trace context in request (or message) headers, server parses these headers and restores the context. For an example see https://opentelemetry.io/docs/languages/java/instrumentation/#context-propagation-between-http-requests
I don't know much about websocket. It is possible that instrumentation can not be done for websockets because it is a low level protocol. You may need to instrument the higher level protocol that you speak over websockets and introduce the tracing headers there.

**Reply by yatamraghu7** [reply]
Hi @laurit,

I have added some code changes and tracepropagators are working fine with B3 headers. I wanted to achieve below

```
Currently my logging pattern is a below

2024-09-13 16:35:46.725  INFO 75976 --- [nio-8080-exec-1] i.o.e.logging.LoggingSpanExporter        : 'GET' : 01011111001110102222222222222222 5fa635b7db8ac85e CLIENT [tracer: io.opentelemetry.spring-web-3.1:2.7.0-alpha] AttributesMap{data={http.request.method=GET, server.address=localhost, http.response.status_code=200, url.full=http://localhost:8081/hello/fromRestTemplate, server.port=8081}, capacity=128, totalAddedValues=5}


2024-09-13 16:35:46.742  INFO 75976 --- [nio-8080-exec-1] i.o.e.logging.LoggingSpanExporter        : 'GET /consume' : 01011111001110102222222222222222 1a12f47bd3c55b80 SERVER [tracer: io.opentelemetry.spring-webmvc-5.3:2.7.0-alpha] AttributesMap{data={network.protocol.version=1.1, http.route=/consume, http.request.method=GET, url.path=/consume, server.address=localhost, client.address=127.0.0.1, user_agent.original=curl/7.88.1, network.peer.address=127.0.0.1, url.scheme=http, http.response.status_code=200, server.port=8080, network.peer.port=33762}, capacity=128, totalAddedValues=12}
```

After adding logging pattern to my application.properties as below

`logging.pattern.level=%5p [${spring.application.name:},traceId=%X{traceId:-},spanId=%X{spanId:-}] %d{yyyy-MM-dd HH:mm:ss.SSS} %logger{36} - %msg%n`

I am getting below log format

```
2024-09-13 16:37:18.030  INFO [Consumer,traceId=,spanId=] 2024-09-13 16:37:18.030 i.o.e.logging.LoggingSpanExporter - 'GET /consume' : 01011111001110102222222222222222 c21a735037e2eb4d SERVER [tracer: io.opentelemetry.spring-webmvc-5.3:2.7.0-alpha] AttributesMap{data={url.scheme=http, http.request.method=GET, server.port=8080, network.peer.port=33592, network.protocol.version=1.1, http.route=/consume, server.address=localhost, client.address=127.0.0.1, user_agent.original=curl/7.88.1, http.response.status_code=200, network.peer.address=127.0.0.1, url.path=/consume}, capacity=128, totalAddedValues=12}
 77429 --- [nio-8080-exec-1] i.o.e.logging.LoggingSpanExporter        : 'GET /consume' : 01011111001110102222222222222222 c21a735037e2eb4d SERVER [tracer: io.opentelemetry.spring-webmvc-5.3:2.7.0-alpha] AttributesMap{data={url.scheme=http, http.request.method=GET, server.port=8080, network.peer.port=33592, network.protocol.version=1.1, http.route=/consume, server.address=localhost, client.address=127.0.0.1, user_agent.original=curl/7.88.1, http.response.status_code=200, network.peer.address=127.0.0.1, url.path=/consume}, capacity=128, totalAddedValues=12}
```


I wanted to achieve below

```
2024-09-13 16:37:18.030  INFO **[Consumer,traceId=01011111001110102222222222222222,spanId=c21a735037e2eb4d]** 2024-09-13 16:37:18.030 i.o.e.logging.LoggingSpanExporter - 'GET /consume' : 01011111001110102222222222222222 c21a735037e2eb4d SERVER [tracer: io.opentelemetry.spring-webmvc-5.3:2.7.0-alpha] AttributesMap{data={url.scheme=http, http.request.method=GET, server.port=8080, network.peer.port=33592, network.protocol.version=1.1, http.route=/consume, server.address=localhost, client.address=127.0.0.1, user_agent.original=curl/7.88.1, http.response.status_code=200, network.peer.address=127.0.0.1, url.path=/consume}, capacity=128, totalAddedValues=12}
 77429 --- [nio-8080-exec-1] i.o.e.logging.LoggingSpanExporter        : 'GET /consume' : 01011111001110102222222222222222 c21a735037e2eb4d SERVER [tracer: io.opentelemetry.spring-webmvc-5.3:2.7.0-alpha] AttributesMap{data={url.scheme=http, http.request.method=GET, server.port=8080, network.peer.port=33592, network.protocol.version=1.1, http.route=/consume, server.address=localhost, client.address=127.0.0.1, user_agent.original=curl/7.88.1, http.response.status_code=200, network.peer.address=127.0.0.1, url.path=/consume}, capacity=128, totalAddedValues=12}
```

Can you please let me know how can i change the logging pattern using application.property and get the traceId and spanId for every line irrespective of my logging level i.e. INFO or DEBUG or TRACE or ERROR.

This will help a lot for our project structure to get our logging and instrumentation in a good shape.

Thanks

**Reply by laurit** [reply]
Strictly speaking what you are asking does not make sense. Exporting a span is a background activity performed by the agent/opentelemetry sdk there is no active trace during that operation. What you should focus on are your own log statements that are emitted while there is an active trace. With the agent this should work out of the box. With spring starter I have not tried myself, but I suspect that this might not work out of the box and you may need follow https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/log4j/log4j-context-data/log4j-context-data-2.17/library-autoconfigure/README.md or https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/logback/logback-mdc-1.0/library/README.md to get this working. @jeanbisutti @zeitlinger could you confirm whether spring starter supports adding traceId/spanId to logs.

**Reply by yatamraghu7** [reply]
Hi @laurit @jeanbisutti @zeitlinger 

Sorry that my question was little bit confusing as I have provided a log statement from my sample application which is having logexporter in the log. I meant I would like to have my log formatted to the below format

2024-09-13 16:37:18.030  INFO **[Consumer,traceId=01011111001110102222222222222222,spanId=c21a735037e2eb4d]** 

As the current format looks like below

2024-09-13 16:35:46.725  INFO 75976 --- [nio-8080-exec-1]    : 'GET' : 01011111001110102222222222222222 5fa635b7db8ac85e CLIENT [tracer: io.opentelemetry.spring-web-3.1:2.7.0-alpha] AttributesMap{data={http.request.method=GET, server.address=localhost, http.response.status_code=200, url.full=http://localhost:8081/hello/fromRestTemplate, server.port=8081}, capacity=128, totalAddedValues=5}

Just ignore the exporter statement in the log. What I am asking is can i somehow get the traceId and spanId appended to each and every line of the log that my logger is writing ?

I would like to achieve this with may be spring boot application property as mentioned below which is not a working example but

logging.pattern.level=%5p [${spring.application.name:},traceId=%X{traceId:-},spanId=%X{spanId:-}] %d{yyyy-MM-dd HH:mm:ss.SSS} %logger{36} - %msg%n

Kindly let me know if this can be achieved without making a logback.xml or log4j.xml

Please checkout my sample code at below repo
https://github.com/yatamraghu7/ManualInstrumentationOtel

![Screenshot from 2024-09-16 11-37-53](https://github.com/user-attachments/assets/1c1ff555-fe7b-4a0f-bb25-7cbc283c08d3)


Thanks

**Reply by yatamraghu7** [reply]
Hi @jkwatson 

I can see that you have suggested some pom changes here https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/3074 and it worked for him to get the traceId and spanId
I am also facing the same issue here that my traceId and spanId that was propagated by opentelemetry was not injected to MDC.
Please check my code at https://github.com/yatamraghu7/ManualInstrumentationOtel
In my logs below I am getting null value can you please suggest where is the mistake happening in my application 

![Screenshot from 2024-09-16 13-50-52](https://github.com/user-attachments/assets/5405f5fc-7cb5-4ccc-a8a6-3408cfedefa3)


You can see in my above comments I wanted to achieve something similar to what was mentioned [https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/3074](url)

Thanks

**Reply by yatamraghu7** [reply]
Hi @laurit and @jkwatson 

I have added the logback.xml file and also mvn dependency below

```
                <dependency>
			<groupId>io.opentelemetry.instrumentation</groupId>
			<artifactId>opentelemetry-logback-mdc-1.0</artifactId>
			<version>2.7.0-alpha</version>
		</dependency>
```

I am still not able to understand why the traceId and spanId are not getting included in MDC from opentelemetry ?

As you can see below screen shot I am missing the traceId and spanId from MDC and also the response from SERVER side also not appending the traceId and spanId to the log statement

```
2024-09-16 15:56:14 - c.m.c.C.C.ConsumerController - INFO  - The retrieved MDC traceId and spanId will be [null] and [null]
				trace_id:01011111001110102222222222222222 span_id:65be5262ffc3f3b3 
2024-09-16 15:56:14 - c.m.c.C.C.ConsumerController - INFO  - This is a call going to happen from consume method to producer class controller method i.e. sayHelloUsingWebClient
				trace_id:01011111001110102222222222222222 span_id:65be5262ffc3f3b3 
2024-09-16 15:56:14 - i.o.e.logging.LoggingSpanExporter - INFO  - 'GET' : 01011111001110102222222222222222 d2d44619fd5ecaa8 CLIENT [tracer: io.opentelemetry.spring-webflux-5.3:2.7.0-alpha] AttributesMap{data={server.address=localhost, server.port=8081, http.request.method=GET, http.response.status_code=200, url.full=http://localhost:8081/hello/fromWebClient}, capacity=128, totalAddedValues=5}
				trace_id:01011111001110102222222222222222 span_id:d2d44619fd5ecaa8 
2024-09-16 15:56:14 - i.o.e.logging.LoggingSpanExporter - INFO  - 'GET /consume/usingWebClient' : 01011111001110102222222222222222 65be5262ffc3f3b3 SERVER [tracer: io.opentelemetry.spring-webmvc-5.3:2.7.0-alpha] AttributesMap{data={network.peer.address=127.0.0.1, server.address=localhost, client.address=127.0.0.1, user_agent.original=curl/7.88.1, url.path=/consume/usingWebClient, server.port=8080, network.peer.port=50532, http.request.method=GET, http.response.status_code=200, http.route=/consume/usingWebClient, network.protocol.version=1.1, url.scheme=http}, capacity=128, totalAddedValues=12}
				trace_id: span_id: 
```
![Screenshot from 2024-09-16 15-58-28](https://github.com/user-attachments/assets/a19e6db5-b281-4469-b2c1-178548f807da)


Can someone please tell me why this is happening.
Sample code can be checked at here https://github.com/yatamraghu7/ManualInstrumentationOtel

Thanks

**Reply by laurit** [reply, marked as answer]
> As you can see below screen shot I am missing the traceId and spanId from MDC 

You won't be able to get the trace id with `MDC.get("trace_id")`. Have a look at the instrumentation in https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/logback/logback-mdc-1.0/library You'll see that the trace id isn't really added to the MDC but to the copy of MDC that is attached to the logging event.

> and also the response from SERVER side also not appending the traceId and spanId to the log statement

I have no idea what you mean by that. If you mean that it is missing from stuff logged by `LoggingSpanExporter` then that is expected.
Overall to me it seems that trace_id and span_id are added when you log from your own code, so seems to be working fine.

**Reply by yatamraghu7** [reply]
Hi @laurit 

Is there any way that I can customize the pattern of the logback.xml file like below 

Current:

%d{HH:mm:ss.SSS} trace_id=%X{trace_id} span_id=%X{span_id} trace_flags=%X{trace_flags} %msg%n

Wanted to Have like:

%d{yyyy-MM-dd HH:mm:ss.SSS} [%thread] [%X{trace_id}, %X{span_id}] %-5level %logger{36} - %msg%n

Is this even possible that we can customize the logging pattern as we like ?

Thanks

**Reply by laurit** [reply]
Could you elaborate what is preventing you from changing the pattern in logback.xml? Whatever it currently has was added by you.

**Reply by yatamraghu7** [reply]
Hi @laurit 

It is working fine and I am able to set the custom logging pattern.

Thanks

**Comment by yatamraghu7** [comment]
I got the resolution from @laurit and thanks for support.
